### DISCUSSION TITLE
Making aggregate_builder available separately

### DISCUSSION THREAD

**Discussion by milgner** [timescale/timescaledb-toolkit#682] [Q&A] [answered]
I was looking at writing a new aggregate function using `pgx` and in the process stumbled over the `aggregate_builder` crate inside the TimescaleDB toolkit. At first I thought that I could just put

```
aggregate_builder = { git = "https://github.com/timescale/timescaledb-toolkit.git", ref = "1.13.1" }
```

into my `Cargo.toml` and use it, but then it turned out that the generated code uses a couple of non-public utility functions from the main crate.

I think it would be really neat to have this available separately. What do you think about extracting the utility functions into their own crate, too, so they could be referenced and then making `aggregate_builder` available as a separate crate?

**Comment by epgts** [comment, marked as answer]
Hi Marcus!

Thanks for the suggestion.

This crate was an experiment that helped us a little but doesn't quite do what we want.  It is, of course, available to you under the terms of the [Timescale License](https://github.com/timescale/timescaledb/blob/master/tsl/LICENSE-TIMESCALE), but we aren't going to be doing any work on it to make it easier to use outside toolkit.

The timing of your question is interesting, as I have been experimenting with another idea to help us build safer postgresql extensions.  It, too, will be an early experiment and not something we are prepared to support outside of toolkit for the foreseeable future, but further down the road, who knows.

Thanks!

**Reply by milgner** [reply]
That's great to hear!
In the meantime I conducted some experiments with `pgx` and its aggregate functionality and found that it already covered the very basic usecase at hand very nicely.
But for some reason the timescaledb-toolkit source code had ranked further up in my search for Rust-based custom aggregates 😅
